logic/src/commonMain/kotlin/com/wire/kalium/logic/network/SessionManagerImpl.kt

```diff
@@ -88,42 +89,37 @@ class SessionManagerImpl internal constructor(
             .onSuccess { serverConfig = it }
             .fold({ error("use serverConfig is missing or an error while reading local storage") }, { it })
         serverConfig!!
-    }!!
+    }
 
-    override suspend fun updateLoginSession(newAccessTokenDTO: AccessTokenDTO, newRefreshTokenDTO: RefreshTokenDTO?): SessionDTO? =
-        wrapStorageRequest {
-            tokenStorage.updateToken(
-                userId = userId.toDao(),
-                accessToken = newAccessTokenDTO.value,
-                tokenType = newAccessTokenDTO.tokenType,
-                refreshToken = newRefreshTokenDTO?.value
+    override suspend fun updateToken(
+        accessTokenApi: AccessTokenApi,
+        oldAccessToken: String,
+        oldRefreshToken: String
+    ): SessionDTO {
+        val refresher = accessTokenRefresherFactory.create(accessTokenApi)
+        return withContext(coroutineContext) {
+            refresher.refreshTokenAndPersistSession(oldRefreshToken).onFailure {
+                if (it is NetworkFailure.ServerMiscommunication) {
+                    onServerMissCommunication(it)
+                }
+            }
+        }.map { refreshResult ->
+            SessionDTO(
+                userId = userId.toApi(),
+                tokenType = refreshResult.accessToken.tokenType,
+                accessToken = refreshResult.accessToken.value,
+                refreshToken = refreshResult.refreshToken.value,
+                cookieLabel = null
             )
-        }.map {
-            sessionMapper.fromEntityToSessionDTO(it)
-        }.onSuccess {
-            session = it
```

Comment:
Where do we assign the new session to `session`? Or should this property be removed?

Reply:
Nice catch.

I removed it during debugging and forgot to add again.

This code also had no tests before, and I didn't cover every single thing yet.

I'll fix it and cover this one too.

Makes me want to transform `session` into an observable, instead of whatever this this.

Reply:
Fixed and covered with tests